File: 프로그래머스/2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.cpp
#include <string>
#include <vector>
#include <queue>

using namespace std;

int solution(int bridge_length, int weight, vector<int> truck_weights) {
    int time = 0;
    int nowWeight = 0;
    queue<int> bridge;
    queue<int> wait;
    
    while(bridge.size() < bridge_length)
        bridge.push(0);
    
    for (int& i : truck_weights) {
        wait.push(i);
    }
    
    while(!bridge.empty()) {
        nowWeight -= bridge.front(); bridge.pop();
        
        if (!wait.empty()) {
            if (nowWeight+wait.front() <= weight) {
                bridge.push(wait.front()); 
                nowWeight += wait.front();
                wait.pop();
            }
            else
                bridge.push(0);
        }
        time++;
    }
    
    return time;
}

File: 프로그래머스/2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.py
from collections import deque

def solution(bridge_length, weight, truck_weights):
    answer = 0
    bridge = deque()
    trucks = deque(truck_weights)
    nowWeight = 0
    
    for i in range (0, bridge_length):
        bridge.append(0)
    
    while len(bridge) != 0 or len(trucks) != 0:
        nowWeight -= bridge.popleft()
        
        if len(trucks) > 0:
            if len(bridge) < bridge_length and nowWeight + trucks[0] <= weight:
                bridge.append(trucks[0])
                nowWeight += trucks.popleft()
            else:
                bridge.append(0)
                
        answer += 1
    
    return answer